### PR TITLE
 Introduce CrossGen Comparison scenario

### PR DESCRIPTION
**Pull request by echesakovMSFT**
This introduces a new CI scenario that will be run against each PR and does the following:
1. Build Linux/arm coreclr under Docker on Linux/x64
2. Generate layout only in Core_Root
3. Run x86_arm crossgen on System.Private.CoreLib.dll and all assemblies in Core_Root
4. Collect information about x86_arm crossgen behavior and output such as return code, stdout/stderr, sha256 hash of the resulting file (this is done via tests/scripts/crossgen_comparison.py)
5. Copy the collected information and some binaries to Linux/arm machine
6. Run arm_arm crossgen on the same set of assemblies and collect all the information as in 4
7. Do **comparison** of behaviors of x86_arm and arm_arm crossgens

**NOTE:** This job is restricted to Ubuntu arm Checked which will be relaxed later to other cross-compilation scenarios. `Microsoft.VisualBasic.dll` assembly is excluded due to https://github.com/dotnet/coreclr/issues/19370

The changes has been tested in #19455

/cc @dotnet/jit-contrib

### CODE REVIEW

File: tests/scripts/crossgen_comparison.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
+#!  /usr/bin/env python
```
Comment (BruceForstall) [inline tests/scripts/crossgen_comparison.py]: In general, I'd like to see more comments. Especially about the scenario this is used in, and the arguments that are passed to it. Maybe with an example.

File: tests/scripts/crossgen_comparison.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+        args.append(out_filename)
+        args.append(in_filename)
+        return args
+
+def compute_file_hashsum(filename):
+    algo=hashlib.sha256()
+    block_size_in_bytes = 65536
+    with open(filename, 'rb') as file:
+        while True:
+            block = file.read(block_size_in_bytes)
```
Comment (BruceForstall) [inline tests/scripts/crossgen_comparison.py]: Q: Does this read "up to block_size_in_bytes"? Wouldn't want it to fail or skip the data if it was less.

File: tests/scripts/crossgen_comparison.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        else:
+            print('{0}  [MISMATCH]'.format(assembly_name.ljust(column_width)))
+            has_mismatch_error = True
+
+    for assembly_name in sorted(base_assemblies - diff_assemblies):
+        print('{0}  [BASE ONLY]'.format(assembly_name.ljust(column_width)))
+        has_mismatch_error = True
+
+    for assembly_name in sorted(diff_assemblies - base_assemblies):
+        print('{0}  [DIFF ONLY]'.format(assembly_name.ljust(column_width)))
```
Comment (BruceForstall) [inline tests/scripts/crossgen_comparison.py]: Have you seen all the different success/error cases in your testing?

File: netci.groovy [hunk 4]
```diff
@@ -425,7 +438,8 @@ class Constants {
                'gcstress0xc_zapdisable_heapverify1',
                'gcstress0xc_jitstress1',
                'gcstress0xc_jitstress2',
-               'gcstress0xc_minopts_heapverify1'
+               'gcstress0xc_minopts_heapverify1',
+               'crossgen_comparison'
```
Comment (BruceForstall) [inline netci.groovy]: this list is sorted identically to the "basicScenarios" + etc. Thus, this should be added after corefx_innerloop, which is missing (!), thus, after "illink".

File: netci.groovy [hunk 5]
```diff
@@ -3589,6 +3692,9 @@ def static shouldGenerateFlowJob(def scenario, def isPR, def architecture, def c
             return false
         }
     }
+    else if (isCrossGenComparisonScenario(scenario)) {
+        return shouldGenerateCrossGenComparisonJob(os, architecture, configuration, scenario)
```
Comment (BruceForstall) [inline netci.groovy]: I'd prefer this logic be inlined below, as is done for all other scenarios. Same in the other place this is used. But I'm ok with this way, also.

File: tests/scripts/crossgen_comparison.py (hunk 2 above)
Comment (echesakovMSFT) [reply tests/scripts/crossgen_comparison.py]: Correct. It means at most `block_size_in_bytes` bytes to read.  If the EOF has been reached, `file.read()` will return `None`.

File: tests/scripts/crossgen_comparison.py (hunk 3 above)
Comment (echesakovMSFT) [reply tests/scripts/crossgen_comparison.py]: Yes, when tested locally I have seen all of the cases. 

Actually, my original plan was to commit these changes as a first iteration and then continue working on getting comparison report in the script more informative.

**Comment by echesakovMSFT** [comment]
@dotnet-bot test ci

File: tests/scripts/crossgen_comparison.py (hunk 1 above)
Comment (jashook) [reply tests/scripts/crossgen_comparison.py]: Please add a standard header. A good model is: https://github.com/dotnet/coreclr/blob/master/tests/scripts/lst_creator.py#L1

File: tests/scripts/crossgen_comparison.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+        has_mismatch_error = True
+
+    sys.exit(1 if has_mismatch_error else 0)
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    subparsers = parser.add_subparsers()
+
+    corelib_parser = subparsers.add_parser('crossgen_corelib')
```
Comment (jashook) [inline tests/scripts/crossgen_comparison.py]: Please hoist the argparser out to follow existing scripts, by placing it close to the header. See: https://github.com/dotnet/coreclr/blob/master/tests/scripts/lst_creator.py#L27

**Comment by echesakovMSFT** [comment]
@dotnet-bot test ci

File: tests/scripts/crossgen_comparison.py (hunk 6 above)
Comment (jashook) [reply tests/scripts/crossgen_comparison.py]: More examples: https://github.com/dotnet/coreclr/blob/master/tests/runtest.py#L34

**Comment by echesakovMSFT** [comment]
@dotnet-bot test ci

**Comment by echesakovMSFT** [comment]
Commented/reformatted Python script, added a standard header and an example. 

@jashook @BruceForstall Can you please take another look at the file? Please let me know if I need clarify/add something.

**Comment by echesakovMSFT** [comment]
Rebased/squashed into two final commits

**Comment by echesakovMSFT** [comment]
@dotnet-bot test ci

**Pull request merged**